File: 0151-reverse-words-in-a-string/151.c
void reverseWord(char *s, int start, int end) {
  --end;
  while (start < end) {
    char temp = s[start];
    s[start]  = s[end];
    s[end]    = temp;
    start++;
    end--;
  }
}

char *reverseWords(char *s) {
  int write_pos = 0, i = 0;

  // Remove extra spaces: keep single space between words
  while (s[i]) {
    if (s[i] != ' ' || (i > 0 && s[i - 1] != ' '))
      s[write_pos++] = s[i];
    ++i;
  }

  // Remove trailing space if any
  if (s[write_pos - 1] == ' ')
    write_pos--;
  s[write_pos] = '\0';

  // Reverse each word
  i = 0;
  while (i < write_pos) {
    int j = i;
    while (j < write_pos && s[j] != ' ') ++j;
    reverseWord(s, i, j);
    i = j + 1;
  }

  // Reverse the whole string
  reverseWord(s, 0, write_pos);

  return s;
}

File: 0151-reverse-words-in-a-string/151.rs
impl Solution {
    pub fn reverse_words(s: String) -> String {
        /* Approach:
         *      1. [C-Style Logical Implementation] -- T.C: O(N), S.C: O(N)
         */
        
        // Approach 2: Rust Idiomatic
        s.split_whitespace()
            .rev()
            .collect::<Vec<&str>>()
            .join(" ")

        /* Approach 1: C-Style Logical Implementation
        let mut chars: Vec<char> = s.chars().collect();

        // Remove extra spaces
        let mut write_pos = 0;
        let n = chars.len();
        let mut i = 0;

        while i < n {
            if chars[i] != ' ' || (i > 0 && chars[i - 1] != ' ') {
                chars[write_pos] = chars[i];
                write_pos += 1;
            }
            i += 1;
        }

        if write_pos > 0 && chars[write_pos - 1] == ' ' {
            write_pos -= 1;
        }
        chars.truncate(write_pos);

        // Reverse each word
        let mut i = 0;
        while i < chars.len() {
            let mut j = i;
            while j < chars.len() && chars[j] != ' ' {
                j += 1;
            }
            chars[i..j].reverse();
            i = j + 1;
        }

        // Reverse whole string
        chars.reverse();
        chars.into_iter().collect()
        */
    }
}
